packages/data/src/components/with-dispatch/test/index.js

```diff
@@ -10,224 +11,138 @@ import withDispatch from '../';
 import { createRegistry } from '../../../registry';
 import { RegistryProvider } from '../../registry-provider';
 
-describe( 'withDispatch', () => {
-	let registry;
-	beforeEach( () => {
-		registry = createRegistry();
-	} );
+jest.useRealTimers();
 
-	it( 'passes the relevant data to the component', () => {
-		const store = registry.registerStore( 'counter', {
-			reducer: ( state = 0, action ) => {
-				if ( action.type === 'increment' ) {
-					return state + action.count;
-				}
-				return state;
-			},
-			actions: {
-				increment: ( count = 1 ) => ( { type: 'increment', count } ),
-			},
-		} );
-
-		const Component = withDispatch( ( _dispatch, ownProps ) => {
+describe( 'withDispatch', () => {
+	const storeOptions = {
+		reducer: ( state = 0, action ) => {
+			if ( action.type === 'inc' ) {
+				return state + action.count;
+			}
+			return state;
+		},
+		actions: {
+			increment: ( count = 1 ) => ( { type: 'inc', count } ),
+		},
+		selectors: {
+			getCount: ( state ) => state,
+		},
+	};
+
+	it( 'passes the relevant data to the component', async () => {
+		const user = userEvent.setup();
+		const registry = createRegistry();
+		registry.registerStore( 'counter', storeOptions );
+
+		const Component = withDispatch( ( dispatch, ownProps ) => {
 			const { count } = ownProps;
 
 			return {
 				increment: () => {
-					const actionReturnedFromDispatch = Promise.resolve(
-						_dispatch( 'counter' ).increment( count )
-					);
-					return expect(
-						actionReturnedFromDispatch
-					).resolves.toEqual( {
-						type: 'increment',
-						count,
-					} );
+					dispatch( 'counter' ).increment( count );
 				},
 			};
 		} )( ( props ) => <button onClick={ props.increment } /> );
 
-		let testRenderer;
-		act( () => {
-			testRenderer = TestRenderer.create(
-				<RegistryProvider value={ registry }>
-					<Component count={ 0 } />
-				</RegistryProvider>
-			);
-		} );
-		const testInstance = testRenderer.root;
-
-		const incrementBeforeSetProps =
-			testInstance.findByType( 'button' ).props.onClick;
+		const { rerender } = render(
+			<RegistryProvider value={ registry }>
+				<Component count={ 0 } />
+			</RegistryProvider>
+		);
 
 		// Verify that dispatch respects props at the time of being invoked by
 		// changing props after the initial mount.
-		act( () => {
-			testRenderer.update(
-				<RegistryProvider value={ registry }>
-					<Component count={ 2 } />
-				</RegistryProvider>
-			);
-		} );
-
-		// Function value reference should not have changed in props update.
-		expect( testInstance.findByType( 'button' ).props.onClick ).toBe(
-			incrementBeforeSetProps
```

Comment:
This is also a useful property worth testing: that the `increment` prop is always an identical function, even though after a rerender with different props it behaves differently. One way to test this is wrapping the `button` in a helper component with `React.memo` and then checking there are no extra rerenders.

Reply:
Would something like [2fb959a](https://github.com/WordPress/gutenberg/pull/44855/commits/2fb959a838644bca7e704344fd4a9a3cd0780bd8) work?